src/native/libs/System.Native/pal_dynamicload.c

```diff
@@ -50,3 +50,20 @@ void SystemNative_FreeLibrary(void* handle)
 {
     dlclose(handle);
 }
+
+#ifdef TARGET_ANDROID
+void* SystemNative_GetDefaultSearchOrderPseudoHandle(void)
+{
+    return (void*)RTLD_DEFAULT;
+}
+#else
+static void* g_defaultSearchOrderPseudoHandle = NULL;
+void* SystemNative_GetDefaultSearchOrderPseudoHandle(void)
+{
+    if (g_defaultSearchOrderPseudoHandle == NULL)
```

Comment:
This has potential race condition depending on how the C++ decides to compile it. The C/C++ compiler is free to reorder the if check and reading of `g_defaultSearchOrderPseudoHandle` to return. It means that the method can return NULL instead of the actual value.

This is real problem. We had many bugs like that over the years - they are always very hard to trace down.

Reply:
This can be fixed by making `g_defaultSearchOrderPseudoHandle` volatile, correct?

Reply:
Yup. Also, you may want to cache the value in a local so that it does not need to be fetched from memory twice.